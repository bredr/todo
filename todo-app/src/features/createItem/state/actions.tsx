import { ActionType, createAction } from "typesafe-actions";

export const open = createAction("OPEN_CREATE");
export const close = createAction("CLOSE_CREATE");

export type Actions = ActionType<typeof open> | ActionType<typeof close>;

export default { open, close };
