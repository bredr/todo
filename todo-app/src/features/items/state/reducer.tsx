import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Item } from "../../../proto/todo_pb";
import {
  Actions,
  cancelEditItem,
  editItem,
  getItems,
  loadingItems,
  setItems
} from "./actions";

interface IEditableItem extends Item.AsObject {
  edit: boolean;
}

export interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: IEditableItem[];
  readonly loading: boolean;
}

const initState: IState = {
  items: [],
  limit: 100,
  loading: false,
  offset: 0,
  total: 0
};

export default createReducer<IState, Actions>(initState)
  .handleAction(loadingItems, (state, action) =>
    produce(state, draft => {
      draft.loading = true;
      draft.items = [];
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
    })
  )
  .handleAction(setItems, (state, action) =>
    produce(state, draft => {
      draft.loading = false;
      draft.limit = action.payload.limit;
      draft.offset = action.payload.offset;
      draft.total = action.payload.total;
      draft.items = action.payload.itemsList.map(item => ({
        ...item,
        edit: false
      }));
    })
  )
  .handleAction(editItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: true } : item
      );
    })
  )
  .handleAction(cancelEditItem, (state, action) =>
    produce(state, draft => {
      draft.items = state.items.map(item =>
        item.id === action.payload ? { ...item, edit: false } : item
      );
    })
  );
