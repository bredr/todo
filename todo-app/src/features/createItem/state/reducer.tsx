import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, close, open } from "./actions";

export interface IState {
  readonly isOpen: boolean;
}

const initState: IState = {
  isOpen: false
};

export default createReducer<IState, Actions>(initState)
  .handleAction(open, state =>
    produce(state, draft => {
      draft.isOpen = true;
    })
  )
  .handleAction(close, state =>
    produce(state, draft => {
      draft.isOpen = false;
    })
  );
