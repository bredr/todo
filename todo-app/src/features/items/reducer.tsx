import { Item, Items } from "../../proto/todo_pb";
import produce from "immer";
import { createReducer } from "typesafe-actions";
import { setItems, loadingItems, getItems, Actions } from "./actions";

interface IState {
  readonly total: number;
  readonly limit: number;
  readonly offset: number;
  readonly items: Array<Item.AsObject>;
  readonly loading: boolean;
}

const initState: IState = {
  total: 0,
  limit: 100,
  offset: 0,
  items: [],
  loading: false
}

export default createReducer<IState, Actions>(initState).
  handleAction(getItems, (state, action) => produce(state, (draft) => {
    draft.items = [];
    draft.limit = action.payload.request.getLimit();
    draft.offset = action.payload.request.getOffset();
    draft.total = 0;
  })).
  handleAction(loadingItems, (state, _action) => produce(state, (draft) => {
    draft.loading = true;
  })).
  handleAction(setItems, (state, action) => produce(state, (draft) => {
    draft.loading = false;
    draft.limit = action.payload.limit;
    draft.offset = action.payload.offset;
    draft.total = action.payload.total;
    draft.items = action.payload.itemsList;
  }))
