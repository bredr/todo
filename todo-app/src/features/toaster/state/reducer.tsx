import produce from "immer";
import { createReducer } from "typesafe-actions";
import { Actions, addError, addSuccess, clearMessage } from "./actions";

interface IMessage {
  error: boolean;
  message: string;
}
export interface IState {
  readonly messages: IMessage[];
}

const initState: IState = {
  messages: []
};

export default createReducer<IState, Actions>(initState)
  .handleAction(addError, (state, action) =>
    produce(state, draft => {
      draft.messages = [
        { error: true, message: action.payload },
        ...state.messages
      ];
    })
  )
  .handleAction(addSuccess, (state, action) =>
    produce(state, draft => {
      draft.messages = [
        { error: false, message: action.payload },
        ...state.messages
      ];
    })
  )
  .handleAction(clearMessage, state =>
    produce(state, draft => {
      const [, ...remainder] = state.messages;
      draft.messages = remainder;
    })
  );
