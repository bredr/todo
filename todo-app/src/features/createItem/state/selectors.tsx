import { createSelector } from "reselect";
import { IState } from "./reducer";

export const CREATE_SLICE = "CREATE_SLICE";

const sliceSelector = (state: { [CREATE_SLICE]: IState }) =>
  state[CREATE_SLICE];

const createItemOpen = createSelector(
  sliceSelector,
  state => state.isOpen
);

export default { createItemOpen };
