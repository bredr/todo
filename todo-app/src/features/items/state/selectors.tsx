import { createSelector } from "reselect";
import { IState } from "./reducer";

export const ITEMS_SLICE = "ITEMS_SLICE";

const sliceSelector = (state: { [ITEMS_SLICE]: IState }) => state[ITEMS_SLICE];

const items = createSelector(
  sliceSelector,
  state => state.items
);
const loading = createSelector(
  sliceSelector,
  state => state.loading
);
const hasNext = createSelector(
  sliceSelector,
  state => state.total > 0 && state.total - (state.limit + state.offset) > 0
);
const hasPrevious = createSelector(
  sliceSelector,
  state => state.offset > 0 && state.total > 0
);

export default { items, loading, hasNext, hasPrevious };
