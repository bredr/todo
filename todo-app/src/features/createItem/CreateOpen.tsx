import { Fab } from "@material-ui/core";
import { createStyles, makeStyles, Theme } from "@material-ui/core/styles";
import AddIcon from "@material-ui/icons/Add";
import React from "react";
import { useDispatch, useSelector } from "react-redux";
import { open } from "./state/actions";
import selectors from "./state/selectors";

const useStyles = makeStyles((theme: Theme) =>
  createStyles({
    fab: {
      margin: theme.spacing(1)
    }
  })
);

export const CreateOpen = () => {
  const classes = useStyles();
  const dispatch = useDispatch();

  const isOpen = useSelector(selectors.createItemOpen);

  function onClick() {
    dispatch(open());
  }

  return (
    <Fab
      color="primary"
      aria-label="add"
      className={classes.fab}
      onClick={onClick}
      disabled={isOpen}
    >
      <AddIcon />
    </Fab>
  );
};
