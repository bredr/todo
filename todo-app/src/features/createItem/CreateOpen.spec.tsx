import { cleanup, fireEvent, render } from "@testing-library/react";
import React from "react";
import { Provider } from "react-redux";
import { combineReducers, createStore } from "redux";
import { CreateOpen } from "./CreateOpen";
import reducer, { IState } from "./state/reducer";
import { CREATE_SLICE } from "./state/selectors";

const renderWithRedux = (
  ui: JSX.Element,
  {
    initialState,
    store = createStore(
      combineReducers({
        [CREATE_SLICE]: reducer
      }),
      initialState
    )
  }: {
    initialState?: { [CREATE_SLICE]: IState };
    store?: any;
  } = {}
) => {
  return { ...render(<Provider store={store}>{ui}</Provider>) };
};

describe("CreateOpen", () => {
  afterEach(() => cleanup());

  it("renders", () => {
    const { container } = renderWithRedux(<CreateOpen />, {
      initialState: { [CREATE_SLICE]: { isOpen: true } }
    });
    expect(container.firstChild).toMatchSnapshot();
  });

  it("is disabled if isOpen", () => {
    const { getByLabelText } = renderWithRedux(<CreateOpen />, {
      initialState: { [CREATE_SLICE]: { isOpen: true } }
    });
    expect(getByLabelText("add").getAttribute("disabled")).toBe("");
  });
  it("is enabled if not isOpen", () => {
    const { getByLabelText } = renderWithRedux(<CreateOpen />, {
      initialState: { [CREATE_SLICE]: { isOpen: false } }
    });
    expect(getByLabelText("add").getAttribute("disabled")).toBe(null);
  });

  it("on click opens", () => {
    const { getByLabelText } = renderWithRedux(<CreateOpen />, {
      initialState: { [CREATE_SLICE]: { isOpen: false } }
    });
    expect(getByLabelText("add").getAttribute("disabled")).toBe(null);
    fireEvent.click(getByLabelText("add"));
    expect(getByLabelText("add").getAttribute("disabled")).toBe("");
  });
});
