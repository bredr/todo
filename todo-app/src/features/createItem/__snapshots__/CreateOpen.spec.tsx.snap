// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateOpen renders 1`] = `
<button
  aria-label="add"
  class="MuiButtonBase-root MuiFab-root makeStyles-fab-1 MuiFab-primary Mui-disabled Mui-disabled"
  disabled=""
  tabindex="-1"
  type="button"
>
  <span
    class="MuiFab-label"
  >
    <svg
      aria-hidden="true"
      class="MuiSvgIcon-root"
      focusable="false"
      role="presentation"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
      />
    </svg>
  </span>
</button>
`;
